obspy/clients/base.py

```diff
@@ -50,31 +50,198 @@ def get_stations(self, network, station, location, channel, starttime,
+        BaseClient.__init__(self, debug=debug)
+
     @abstractmethod
     def get_service_version(self):
-        """Return a semantic version number as a string."""
+        """
+        Return a semantic version number of the remote service as a string.
+        """
         pass
 
 
+class HTTPClient(with_metaclass(ABCMeta, RemoteBaseClient)):
+    """
+    Mix-in class to add HTTP capabilities.
+
+    :param debug: Passed on to the :class:`BaseClient` constructor.
+    :type debug: bool
+    :param timeout: Passed on to the :class:`RemoteBaseClient` constructor.
+    :type timeout: float
+
+    .. rubric:: Example
+
+    from obspy.clients.base import (WaveformClient, HTTPClient,
+                                    DEFAULT_USER_AGENT)
+
+    class NewClient(WaveformClient, HTTPClient):
+        def __init__(self, user_agent=DEFAULT_USER_AGENT, debug=False,
+                     timeout=20):
+            HTTPClient.__init__(self, user_agent=user_agent, debug=debug,
+                                timeout=timeout)
+
+        def _handle_requests_http_error(self, r):
+            r.raise_for_status()
+
+        def get_service_version(self):
+            ...
+
+        def get_waveforms(...):
+            ...
+    """
+    def __init__(self, debug=False, timeout=120,
+                 user_agent=DEFAULT_USER_AGENT):
+        self._user_agent = user_agent
+        RemoteBaseClient.__init__(self, debug=debug, timeout=timeout)
+
+    @abstractmethod
+    def _handle_requests_http_error(self, r):
+        """
+        Error handling for the HTTP errors.
+
+        Method called when the _download() method downloads something with a
+        status code different than 200.
+
+        The error codes mean different things for different web services
+        thus this needs to be implemented by every HTTPClient.
+
+        :param r: The response object resulting in the error.
+        :type r: :class:`requests.Response`
+        """
+        pass
+
+    def _download(self, url, params=None, filename=None, data=None):
+        """
+        Download the URL with GET or POST and the chosen parameters.
+
+        Will only return if it manages to download with HTTP code 200,
+        otherwise the _handle_requests_http_error() method is called.
+
+        By default it will send a GET request - if data is given it will
+        send a POST request.
+
+        :param url: The URL to download from.
+        :type url: str
+        :param params: Additional URL parameters.
+        :type params: dict
+        :param filename: String or file like object. Will download directly
+            to the file. If given, this function will return nothing.
+        :type filename: str or file-like object
+        :param data: If given, a POST request will be sent with the data in
+            the body of the request.
+        :type data: dictionary, bytes, or file-like object
+        :return: The response object resulting in the error.
```

Comment:
This statement is the opposite of the second paragraph, which says it will only return if the code is 200, i.e., not an error.

In fact, I think whether it returns anything or not has nothing to do with the return code, but whether a location to save the result was specified.